sg_otio/track_diff.py

```diff
@@ -0,0 +1,395 @@
+        if old_track:
+            for i, clip in enumerate(old_track.each_clip()):
+                # Check if we have some SG meta data
+                sg_cut_item = clip.metadata.get("sg")
+                if not sg_cut_item or sg_cut_item.get("type") != "CutItem":
+                    raise ValueError(
+                        "Invalid clip %s not linked to a SG CutItem" % clip.name
+                    )
+                sg_shot = sg_cut_item.get("shot")
+                if sg_shot:
+                    shot_id = sg_shot.get("id")
+                    if shot_id:
+                        sg_shot_ids.append(shot_id)
+                prev_clip_list.append(
+                    SGCutClip(clip, index=i + 1, sg_shot=sg_shot)
+                )
+            logger.debug(
+                "Previous Cut list is %s" % ",".join(
+                    ["%s (%s)" % (x.name, x.sg_shot) for x in prev_clip_list]
+                )
+            )
+        sg_shots_dict = {}
+        # Retrieve details for Shots linked to the Clips
+        if sg_shot_ids:
+            sg_shots = self._sg.find(
+                "Shot",
+                [["id", "in", list(sg_shot_ids)]],
+                sg_shot_fields
+            )
+            # Build a dictionary where Shot names are the keys, use the Shot id
+            # if the name is not set
+            sg_shots_dict = dict(((x["code"] or str(x["id"])).lower(), x) for x in sg_shots)
+
+        # Retrieve additional Shots from the new track if needed
+        self._diffs_by_shots = {}
+        more_shot_names = set()
+        for i, clip in enumerate(new_track.each_clip()):
+            shot_name = compute_clip_shot_name(clip)
+            if shot_name:
+                # Matching Shots must be case insensitive
+                shot_name = shot_name.lower()
+                more_shot_names.add(shot_name)
+            # Ensure a SGCutDiffGroup and add SGCutDiff to it.
+            if shot_name not in self._diffs_by_shots:
+                self._diffs_by_shots[shot_name] = SGCutDiffGroup(shot_name)
+            self._diffs_by_shots[shot_name].add_clip(
+                SGCutDiff(clip=clip, index=i + 1, sg_shot=None)
+            )
+        if more_shot_names:
+            sg_more_shots = self._sg.find(
+                "Shot",
+                [
+                    ["project", "is", self._sg_project],
+                    ["code", "in", list(more_shot_names)]
+                ],
+                sg_shot_fields,
+            )
+            for sg_shot in sg_more_shots:
+                shot_name = sg_shot["code"].lower()
+                self._diffs_by_shots[shot_name].sg_shot = sg_shot
+
+        # Duplicate the list of shots, allowing us to know easily which ones are
+        # not part of the new track by removing entries when we use them.
+        # We only need a shallow copy here
+        leftover_shots = [x for x in sg_shots_dict.values()]
+        seen_names = []
+        duplicate_names = {}
```

Comment:
maybe use a `defaultdict(int)`?

Reply:
will keep it like this if you don't mind.